### PR TITLE
Fix nvlink warning SM Arch ('sm_35') not found

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes the warnings in #2383 similar to
```
nvlink warning : SM Arch ('sm_35') not found in ...
```
On the other hand, this results in warnings like
```
nvlink warning : Stack size for entry function '_ZN6Kokkos4Impl72_GLOBAL__N__48_tmpxft_0000578e_00000000_6_TestCuda_Task_cpp1_ii_c664630423cuda_task_queue_executeINS_19SimpleTaskSchedulerINS_4CudaENS0_15SingleTaskQueueIS4_NS_12CudaUVMSpaceENS0_24TaskQueueTraitsLockBasedENS_10MemoryPoolINS_6DeviceIS4_S6_EEEEEEEEEEvT_i' cannot be statically determined
```
as also reported in #2039.

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
@jjellio , the stack size warning is the one you're seeing. I don't really have time to check this out today, but you might want to know about this issue

**Comment by jjellio** [comment]
So, I read this correctly, the stack warning message popped up because the arch flag was passed to the linker w/relocatable code specified?

It still seems nvcc should have whined about this at compile time as well.

File: .jenkins [hunk 1]
```diff
@@ -95,6 +95,7 @@ pipeline {
                                 -DKokkos_ENABLE_CUDA=ON \
                                 -DKokkos_ENABLE_CUDA_LAMBDA=ON \
                                 -DKokkos_ENABLE_CUDA_UVM=ON \
+                                -DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=ON \
```
Comment (crtrott) [inline .jenkins]: If we do this: rename the test to include "-RDC" or so.

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -388,6 +388,7 @@ IF(KOKKOS_ARCH_${ARCH})
   ENDIF()
   SET(CUDA_ARCH_ALREADY_SPECIFIED ${ARCH} PARENT_SCOPE)
   GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=${FLAG}")
+  GLOBAL_APPEND(KOKKOS_LINK_OPTIONS "${CUDA_ARCH_FLAG}=${FLAG}")
```
Comment (crtrott) [inline cmake/kokkos_arch.cmake]: jup this is required if RDC is on. It shouldn't hurt if it is not.

**Review by crtrott — CHANGES_REQUESTED** [review]
Change the name of the test which turns RDC on to have RDC in its name.

**Comment by crtrott** [comment]
@jjellio this warning is a bit annoying. Basically it says: you are calling a function from a different compilation unit in a device kernel so I can't figure out how many registers that kernel is using. What it doesn't say is that because of it it will just grant it enough registers so that no matter which function in the executable you call, it will work. This is a warning which will always appear, if you needed relocatable device code in the first place ...

**Comment by jjellio** [comment]
We get this in EMPIRE at *compile time* for functions w/recursive stuff.  (no relocatable device code turned on).

I suspect it generates the same code path though - this one simply gives it at link time, but the outcome is probably the same as Christian describes.

It would be nice if Nvidia would support ELF sections for functions/data (e.g., -ffunction-sections -fdata-sections) - then we could garbage collect unused symbols in a static build. (e.g., -Wl,--gc-sections)  (which helps identify poor code organization in the app code)

**Review by crtrott — APPROVED** [review]

**Pull request merged**